_posts/2017-01-09-Jak-psat-php-Zavislosti+verzovaci-systemy+debugging.md

```diff
@@ -0,0 +1,38 @@
+
+# Verzovací systém (Version Control System, zkráceně verzování nebo VCS)
+Jsou mezi námi dlouho a používá je míň lidí, než by bylo vhodné. Verzovací systémy slouží ke snadné správě kódu v čase. Můžete se podívat, kdy jaká změna v kódu nastala a který vývojář ji provedl. Umožňují jednoduše provést "rollback" (návrat v čase) nebo souběžný vývoj více různých větví programu a jejich následné sloučení do jednoho finálního projektu. Při vyvíjení v týmu větším než tři lidi už jsou prakticky nepostradatelné. 
+ 
+Verzovací systém ze své podstaty poskytuje nástroje a výhody zálohování. Ve zdrojových kódech nemusí být zakomentované bloky kódu, které využijeme "někdy příště"; z verzování se v případě potřeby dají snadno vytáhnout z historie. Pomocí verzovacích systému se dá i jednoduše zálohovat i jinam než na vlastní disk, o tom ale více příště. 
+   
+Nejpoužívanějším verzovacím systémem je git (pozor, neplést s <github.com>, který git používá, je to ale cloudový správce kódu>). Git vymyslel a napsal [Linus Torvalds](https://cs.wikipedia.org/wiki/Linus_Torvalds), když vyvíjel Linux. Používá se pro správu takových projektů jako jsou linuxové jádro nebo programovací jazyk Ruby. Je zpočátku složitější na pochopení, ale jeho užitečnost daleko přesahuje počáteční náklady na nastudování. Doporučuju ho používat v podstatě na všechny projekty (mnohdy i na ty neprogramátorské, kde je potřeba pamatovat si historii, kupříkladu diplomek či jiných rozsáhlých textů). Mezi dalšími verzovacími systémy se používají [Subversion](https://subversion.apache.org/) od [Apache](https://www.apache.org/) a [Mercurial](https://www.mercurial-scm.org/) - práce s nimi je podobná jako s gitem, ale žádný jsem zatím nepoužíval a tak je nemůžu doporučit ani od nich odradit. 
+
+Téměř každé IDE umožňuje nějaký stupeň integrace verzovacích systémů, neškodí se ale základy naučit přímo v příkazovém řádku a tedy univerzálně. Dobré návody a rozcestníky na git najdete třeba [tady](https://rogerdudler.github.io/git-guide/), [tady](https://stackoverflow.com/questions/315911/git-for-beginners-the-definitive-practical-guide) nebo [tady](https://www.sitepoint.com/git-for-beginners/). 
+
+# Testování a debugging
+Je dobrým zvykem mít kód otestovaný lépe než použitím `echo($foo)` a k tomu ručním proklikáním celého projektu v prohlížeči. To jsou první a intuitivní praktiky, kterými každý z nás prošel a používal, nesmíme na nich ale zamrznout! 
+
+Pro testování projektu se používají automatizované testy. Je třeba tyto testy napsat a je třeba mít aplikaci, která testy provádí. Testování se většinou dělí na Unit Testing (jednotkové testování), který testuje jednotlivé třidy nebo jiné malé celky a Behaviour/Acceptance Testing (testy chování), který testuje chování aplikace jako celku. 
```

Comment:
Pojmy jako Unit testing a behaviour/acceptance bych sem netahal, to už je spíš pokročilejší.

Psal bych jen "testy"